@here/harp-mapview/lib/VisibleTileSet.ts

```diff
@@ -269,15 +270,29 @@ export class VisibleTileSet {
 
         const newRenderList: DataSourceTileList[] = [];
         let allVisibleTilesLoaded: boolean = true;
+        let allBoundingBoxesFinal: boolean = true;
 
         if (this.options.extendedFrustumCulling) {
             this.m_mapTileCuller.setup();
         }
 
+        let elevationRangeSource: ElevationRangeSource | undefined;
+        for (const dataSource of dataSources) {
+            elevationRangeSource = dataSource.getElevationRangeSource();
+            if (elevationRangeSource !== undefined) {
+                // We don't support multiple elevation range sources, but just take the first one
+                // that we find in the enabled data sources.
+                break;
+            }
+        }
+
         for (const dataSource of dataSources) {
             const displayZoomLevel = dataSource.getDisplayZoomLevel(zoomLevel);
 
             const tilingScheme = dataSource.getTilingScheme();
+            const useElevationRangeSource: boolean =
+                elevationRangeSource !== undefined &&
+                elevationRangeSource.getTilingScheme() === tilingScheme;
```

Comment:
> `elevationRangeSource.getTilingScheme() === tilingScheme`

I tdon't think check belongs here. I would change contract of `DataSource.getElevationRangeSource` that it returns only _compatible_ elevation source.
If there is no compatible elevation source, it should just return undefined.

Reply:
What does a compatible elevation source mean? There is probably only one elevation data source (unless you want to somehow overlay different terrain sets)

Reply:
`getElevationRangeSource` has builtin tiling scheme, so in this case _compatibility_ is defined as "same tiling scheme".

What i mean, that if `DS.getElevationRangeSource` returns something, then it should be usable.
You shouldn't even have incompatible elevation source attached to `DS`.

>  There is probably only one elevation data 

Why?
I understand that there is probably one data set and service that serves elevation data, but each datasource may have different one javascript object instance serving elevation data for different tiling schemes.

> There is probably only one elevation data source

Note, that here we have elevation *range* source, which is only useful for tiling and is heavily bound to tiling scheme.
For general elevation data (i assume) we will need more fine grained interface like `getElevation(lat,long)` - imagine picking or laying out roads on terrain.

...

So, this discussion revolves about question "is elevation range source" property of DS or MapView, so it goes back to this comment: https://github.com/heremaps/harp.gl/pull/366#discussion_r274878015

If it's DS property, then yes, DS shouldn't return incompatible one but on the other hand, we should support many of them

If it's MapView property, then move it to MapView and check tilng scheme here.

Reply:
> > There is probably only one elevation data
> 
> Why?

What use case do you have for two differently enabled tiled terrain sources? What do you want to show the user?

> I understand that there is probably one data set and service that serves elevation data, but each datasource may have different one javascript object instance serving elevation data for different tiling schemes.

The `DataSource` may not know about elevation data, but we know that the terrain data source knows about elevation data, it just extracts the data from the tile and makes it usable for other data sources.

> 
> > There is probably only one elevation data source
> 
> Note, that here we have elevation _range_ source, which is only useful for tiling and is heavily bound to tiling scheme.
> For general elevation data (i assume) we will need more fine grained interface like `getElevation(lat,long)` - imagine picking or laying out roads on terrain.

I think this is a separate topic.

> 
> ...
> 
> So, this discussion revolves about question "is elevation range source" property of DS or MapView, so it goes back to this comment: [#366 (comment)](https://github.com/heremaps/harp.gl/pull/366#discussion_r274878015)
> 
> If it's DS property, then yes, DS shouldn't return incompatible one but on the other hand, we should support many of them

Do you want to force each DS to return a valid `ElevationRangeSource`?

> 
> If it's MapView property, then move it to MapView and check tilng scheme here.

How does the MapView get this data?

Reply:
@zbigg Maybe there is a misunderstanding in the behaviour: The ElevationRangeSource is coming from the terrain and is applied to all data sources that have the same tiling scheme(e.g. OMV but also terrain itself) b/c the will be overlayed on top of terrain. The other data sources don't have an elevation range source. The case that another data source (e.g. OMV) has 3d data and an elevation source is currently not supported.
The reason why I moved the ElevationRangeSource to DataSource and not MapView is that data sources can be disabled. In my first implementation the terrain data source was setting the elevation range source to the MapView, but this is wrong if the terrain data source is disabled.

As mentioned by @nzjony, getting elevation values for arbitrary geo coordinates and ray casting are not part of that change. This will be handled by a different class (e.g. HeightProvider) that will basically use the TerrainTiles that are in the TileCache.

Reply:
Thanks for explanation and sorry for questioning change, it's just adding  interfaces without clear doc & example implementation is confusing and begs for questions.

More implementation and/or use-case for this interfaces would help use understand why it is really needed, but to the point.

After rethinking this again, what i see it that for this particular change - namely, "support culling of tiles with some min/max altitut"e it's enough to add:
```
DataSource::getTileElevationRange(tileKey): ElevationRangeResult | undefined;
```
With basic premise that datasources who does not support elevation data or are not attached to any will just return `undefined`. and be done with it.

Less code, less burden.
(This of course is true, if we don't have any other use case for `ElevationRangeSource` which i'm not aware of yet).

Reply:
How should that work for the data sources that don't have an elevation range for a given key (e.g. OMV)? Should I for these data sources loop through all other data sources and call getTileElevationRange on them?

This all depends on a bigger discussion that did not happen yet, i.e. who is responsible for overlaying vector data:
1. MapView
2. The vector data source itself

I was already tending to go into the direction of 1. but maybe we should have this discussion first (i.e. finish the terrain vector overlay solutioning)